security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,183 @@
+	if err := verifyCSRIdentities(identities, req.GetCsr()); err != nil {
+		return nil, fmt.Errorf("verify CSR Identities failed: %v", err)
+	}
+	cLog.Debugf("forwarding Workload's CSR to Custom CA server...")
+	resp, err := c.signingAPIClient.CreateCertificate(ctx, req)
+	certChain := resp.GetCertChain()
+	if err != nil {
+		cLog.Errorf("cannot call CreateCertificate from Custom CA: %v", err)
+		return nil, fmt.Errorf("cannot call CreateCertificate from Custom CA: %v", err)
+	}
+
+	var responseCertChains []string
+	certChainWithoutRoot := certChain[:len(certChain)-1]
+	for _, cert := range certChainWithoutRoot {
+		parsedCert, err := validateAndParseCert(cert)
+		if err != nil {
+			cLog.Errorf("response certificate from Custom CA is invalid: %v", err)
+			return nil, fmt.Errorf("response certificate from Custom CA is invalid: %v", err)
+		}
+		responseCertChains = append(responseCertChains, parsedCert)
+	}
+
+	// Append current roots: Custom CA's root CA and Self signing's root CA
+	rootCertBytes := c.keyCertBundle.GetRootCertPem()
+	responseCertChains = append(responseCertChains, string(rootCertBytes))
+
+	return &pb.IstioCertificateResponse{
+		CertChain: responseCertChains,
+	}, nil
+}
+
+func validateAndParseCert(cert string) (string, error) {
+	defer responseTimeTrack(time.Now(), "validateAndParseCert")
+	certBytes, _ := pem.Decode([]byte(cert))
+
+	if certBytes == nil {
+		return "", fmt.Errorf("decode cert is failed, invalid certificate: %v", cert)
+	}
+
+	block := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes.Bytes,
+	}
+
+	c := pem.EncodeToMemory(block)
+	return string(c), nil
+}
+
+func verifyCSRIdentities(identities []string, csrPEM string) error {
+	if len(identities) == 0 {
+		return nil
+	}
+
+	csrParsed, _ := pem.Decode([]byte(csrPEM))
+	if csrParsed == nil {
+		return fmt.Errorf("decode csr failed - empty pem: %v", csrPEM)
+	}
+	parsedCSR, err := x509.ParseCertificateRequest(csrParsed.Bytes)
+	if err != nil {
+		return fmt.Errorf("cannot parse CSR: %v", err)
+	}
+
+	if len(parsedCSR.Subject.CommonName) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains Subject.CommonName: %#v", parsedCSR.Subject.CommonName)
+	}
+
+	if len(parsedCSR.DNSNames) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains DNSNames: %v", parsedCSR.DNSNames)
+	}
+
+	if len(parsedCSR.IPAddresses) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains IPAddresses: %v", parsedCSR.IPAddresses)
+	}
+
+	if len(parsedCSR.EmailAddresses) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains EmailAddresses: %v", parsedCSR.EmailAddresses)
+	}
+
+	idsFromCSR, err := util.ExtractIDs(parsedCSR.Extensions)
+	if err != nil {
+		return fmt.Errorf("cannot extract identities from CSR: %v", err)
+	}
+
+	if len(idsFromCSR) == 0 {
+		return fmt.Errorf("missing spiffe identities in CSR")
+	}
+
+	if len(idsFromCSR) > 1 {
+		return fmt.Errorf("received unsupported CSR, contains more than 1 identity: %v", idsFromCSR)
+	}
+
```

Comment:
@howardjohn I checked len of identities

Reply:
Not sure it makes sense to only allow 1 identity (may be wrong!).

I would expect a string match against a sorted list of both the authed identities, and those in the CSR.

Reply:
I think I agree, but I am not sure it matters much for for typical k8s use cases in practice. In these I don't know of any time there are multiple identities. We should probably just match whatever the current logic does - I suspect it puts *all* authed identities, so Josh's suggestion would make sense. Thanks!

Reply:
I did it @howardjohn , multiple CSR's identities is supported now